src/hashtable.h

```diff
@@ -0,0 +1,227 @@
+ *
+ * Example:
+ *
+ *   typedef ... key_t;
+ *   int key_hash(const key_t *e);
+ *   int key_cmp(key_t *e, const key_t *o);
+ *
+ *   typedef struct entry {
+ *     key_t key;  // Inherit from key_t.
+ *     ...extra entry value data...
+ *   } entry_t;
+ *   void entry_init(entry_t *e, ...);
+ *
+ *   hashtable_t *t;
+ *   entry_t entries[300];
+ *   key_t k;
+ *   entry_t *e;
+ *
+ *   t = hashtable_new(300, &key_hash, &key_cmp);
+ *   entry_init(&entries[5], ...);
+ *   hashtable_add(t, &entries[5]);
+ *   k = ...;
+ *   e = hashtable_find(t, &k);
+ *
+ *   hashtable_iter i;
+ *   for (e = hashtable_iter(&i, t); e != NULL; e = hashtable_next(&i))
+ *     ...
+ *
+ *   hashtable_free(t);
+ *
+ * The hash() and cmp() fuctions will typically take pointers to
+ * key/entry instances the same as the pointers stored in the
+ * hashtable. However it is also possible for them to take "match
+ * objects" that are a "subclass" of the entry type that contain
+ * additional state for complicated comparision operations.
+ *
+ * Example:
+ *
+ *   typedef struct match {
+ *     key_t key;  // Inherit from key_t;
+ *     ...extra match criteria and state data...
+ *   } match_t;
+ *   int match_cmp(match_t *m, const entry_t *e);
+ *
+ *   ...
+ *   match_t m;
+ *
+ *   t = hashtable_new(300, &key_hash, &match_cmp);
+ *   ...
+ *   m = ...;
+ *   e = hashtable_find(t, &m);
+ *
+ * The cmp() function is only called for finding hashtable entries
+ * and can mutate the match_t object for doing things like deferred
+ * and cached evaluation of expensive match data. It can also access
+ * the whole entry_t object to match against more than just the key. */
+
+/** The hash() function type.
+ *
+ * Args:
+ *   *k - the key or match object to hash.
+ *
+ * Returns:
+ *   An unsigned hash value.
+ */
+typedef unsigned (*hash_f) (const void *k);
+
+/** The cmp() function type.
+ *
+ * Args:
+ *   *k - the key or match object to try and match to.
+ *   *o - the key or entry object to match against.
+ *
+ * Returns:
+ *   -1, 0, or 1 if *e is less, equal, or more that *o. */
+typedef int (*cmp_f) (void *k, const void *o);
+
+/** The hashtable type. */
+typedef struct _hashtable {
+    int size;                   /* Size of allocated hashtable. */
+    int count;                  /* Number of entries in hashtable. */
+    hash_f hash;                /* Function for hashing entries. */
+    cmp_f cmp;                  /* Function for comparing entries. */
+#ifndef HASHTABLE_NSTATS
+    /* The following are for accumulating hashtable_find() stats. */
+    long find_count;            /* The count of finds tried. */
+    long match_count;           /* The count of matches found. */
+    long keycmp_count;          /* The count of key compares done. */
+    long entrycmp_count;        /* The count of entry compares done. */
+#endif
+    void **etable;              /* Table of pointers to entries. */
+    unsigned ktable[];          /* Table of hash keys. */
```

Comment:
This is true when you have a hash-value match, but it's not true when the hash values are different. If the hash values don't match, you don't need to look at the entry value, so you don't get an extra cache mise.

This is part of where I'm optimizing for lookup misses. For a 716M file with 50% different data, using 1K blocks, for a load factor of 0.7, we get the following match statistics;

librsync: match statistics: signature[375756800 searches, 366592 (0.098%) matches, 1053222017 (2.803x) weak sum compares, 867330 (0.231%) strong sum compares, 866681 (0.231%) strong sum calcs]

Note that there were 1,053,222,017 hash compares (2.803 per search) for 866,681 entry compares (0.231% per search). That's only  0.082% entry compares per hash compare, or one entry compare for every 1215 hash compares.

So by packing the hash values tighter, we get less cache misses for the hash compares, and there is way more of them than entry compares.